### PR TITLE
Testing data fetching via pooch

### PR DESCRIPTION
**Pull request by Zeitsperre**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
  - This PR fixes #xyz
- [ ] (If applicable) Documentation has been added / updated (for bug fixes / features).
- [ ] (If applicable) Tests have been added.
- [x] CHANGES.rst has been updated (with summary of main changes).
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.

### What kind of change does this PR introduce?

* Added `pooch` as a hard dependency (this can be revisited)
* Added `xhydro.testing.helpers.py` and `xhydro.testing.registry.txt`
* Added functions:
    * `generate_registry`: Parses data found in package (`xhydro.testing.data`), and adds it to the `registry.txt`
    * `load_registry`: Loads installed (or custom) registry and returns dictionary
    * `populate_testing_data`: Fetches the registry and optionally caches files at a different location (helpful for `pytest-xdist`)
* Added a pre-commit hook for validating NumPy docstrings, fixed a few docstrings.

### Does this PR introduce a breaking change?

Yes. Testing data management is largely handled at the environment variable level:
* `XHYDRO_DATA_DIR` can be used to override the default cache location set for `pooch` (on *nix: `$XDG_CACHE_HOME`; Must be an absolute path)
* `XHYDRO_TESTDATA_BRANCH` can be used to override the branch that the testing data
should be fetched from (default: `main`).

### Other information:
https://www.fatiando.org/pooch/latest/sample-data.html
Based on a similar approach taken for: https://github.com/Ouranosinc/xclim-testdata

Typically, the loaded `pooch` registries are given dog names, so I landed on `DEVEREAUX` as an appropriate name for this project.

### CODE REVIEW

File: environment-dev.yml [hunk 1]
```diff
@@ -6,6 +6,8 @@ dependencies:
   # Don't forget to sync changes between environment.yml, environment-dev.yml, and pyproject.toml!
   # Main packages
   - numpy
+  - pooch >=1.8.0
+  - pydantic >=2.0,<2.5.3
```
Comment (RondeauG) [inline environment-dev.yml:10]:
```suggestion
  - pydantic >=2.0,<2.5.3  # FIXME: Remove pin once our dependencies (xclim, xscen) support pydantic 2.5.3
```

Comment (RondeauG) [reply environment-dev.yml:10]: I'm not sure if the pin is supposed to be permanent or temporary. If temporary, I'd add this.

File: xhydro/testing/helpers.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+_default_cache_dir = pooch.os_cache("xhydro-testdata")
+
+DATA_DIR = os.getenv("XHYDRO_DATA_DIR", _default_cache_dir)
+"""Sets the directory to store the testing datasets.
+
+If not set, the default location will be used (based on ``platformdirs``, see :func:`pooch.os_cache`).
+
+Notes
+-----
+When running tests locally, this can be set for both `pytest` and `tox` by exporting the variable:
```
Comment (RondeauG) [inline xhydro/testing/helpers.py:32]: I'd add the relevant info in CONTRIBUTING, in the section where we talk about running tests.

**Review by RondeauG — APPROVED** [review]
Hard to really test or review without data in `xhydro-testdata`, but overall this seems good!

File: xhydro/testing/helpers.py (hunk 2 above)
Comment (Zeitsperre) [reply xhydro/testing/helpers.py:32]: Updated. Let me know what you think.

**Review by RondeauG — APPROVED** [review]
LGTM!

**Pull request merged**